File: matlab/+kalman/UKF.m
classdef UKF < handle
%UKF Unscented Kalman filter (MATLAB port of the C kf_ukf_* API).
%
%   Nonlinear models without Jacobians: 2n+1 sigma points are sampled from the
%   Cholesky factor of P, propagated through f/h, and the mean/covariance
%   reconstructed. Parameters: alpha (spread), beta (prior), kappa (scaling).

    properties
        n (1,1) double = 0
        m (1,1) double = 0
        x  double
        P  double
        Q  double
        R  double
        f  function_handle     % f(x, u, dt) -> x_out
        h  function_handle     % h(x) -> z_out
        alpha (1,1) double = 1
        beta  (1,1) double = 2
        kappa (1,1) double = 0
        lambda_ (1,1) double
        Wm double
        Wc double
    end

    methods
        function obj = UKF(n, m, f, h)
            if nargin < 4, h = []; end
            if nargin < 3, f = []; end
            obj.n = n;
            obj.m = m;
            obj.x = zeros(n, 1);
            obj.P = eye(n);
            obj.Q = zeros(n);
            obj.R = zeros(m);
            obj.f = f; obj.h = h;
            % lambda = alpha^2 (n + kappa) - n = 0 for the defaults (matches C)
            obj.lambda_ = obj.alpha^2 * (obj.n + obj.kappa) - obj.n;
            obj = recompute_weights(obj);
        end

        function set_models(obj, f, h)
            obj.f = f; obj.h = h;
        end

        function set_parameters(obj, alpha, beta, kappa)
            if alpha <= 0 || beta < 0
                error('kalman:UKF:badParams', 'alpha > 0 and beta >= 0');
            end
            obj.alpha = alpha; obj.beta = beta; obj.kappa = kappa;
            obj.lambda_ = alpha^2 * (obj.n + kappa) - obj.n;
            obj = recompute_weights(obj);
        end

        function obj = recompute_weights(obj)
            denom = obj.n + obj.lambda_;
            nsig = 2 * obj.n + 1;
            obj.Wm = 0.5 / denom * ones(1, nsig);
            obj.Wc = 0.5 / denom * ones(1, nsig);
            obj.Wm(1) = obj.lambda_ / denom;
            obj.Wc(1) = obj.lambda_ / denom + (1 - obj.alpha^2 + obj.beta);
        end

        function set_state(obj, x), obj.x = double(x(:)); end
        function x = get_state(obj), x = obj.x; end
        function set_covariance(obj, P), obj.P = double(P); end
        function set_covariance_diagonal(obj, d), obj.P = diag(double(d(:))); end
        function set_covariance_scalar(obj, p), obj.P = p * eye(obj.n); end
        function P = get_covariance(obj), P = obj.P; end
        function set_process_noise(obj, Q), obj.Q = double(Q); end
        function set_process_noise_scalar(obj, q), obj.Q = q * eye(obj.n); end
        function set_measurement_noise(obj, R), obj.R = double(R); end
        function set_measurement_noise_scalar(obj, r), obj.R = r * eye(obj.m); end

        function sig = sigma_points(obj)
            L = chol(obj.P)';            % lower; P = L L'
            c = sqrt(obj.n + obj.lambda_);
            sig = zeros(2*obj.n + 1, obj.n);
            sig(1, :) = obj.x';
            for i = 1:obj.n
                sig(i+1, :)     = (obj.x + c * L(:, i))';
                sig(i+1+obj.n, :) = (obj.x - c * L(:, i))';
            end
        end

        function predict(obj, u, dt)
            if nargin < 3, dt = 0; end
            if nargin < 2, u = []; end
            if isempty(obj.f), error('kalman:UKF:noModel', 'call set_models() first'); end
            sig = sigma_points(obj);
            nsig = 2 * obj.n + 1;
            X = zeros(nsig, obj.n);
            for i = 1:nsig
                X(i, :) = obj.f(sig(i, :)', u, dt)';
            end
            xm = obj.Wm * X;                         % 1 x n
            d = X - repmat(xm, nsig, 1);
            obj.P = d' * (obj.Wc' .* d) + obj.Q;
            obj.P = (obj.P + obj.P') / 2;
            obj.x = xm';
        end

        function update(obj, z)
            z = double(z(:));
            if isempty(obj.h), error('kalman:UKF:noModel', 'call set_models() first'); end
            sig = sigma_points(obj);
            nsig = 2 * obj.n + 1;
            Z = zeros(nsig, obj.m);
            for i = 1:nsig
                Z(i, :) = obj.h(sig(i, :)')';
            end
            zm = obj.Wm * Z;                         % 1 x m
            dz = Z - repmat(zm, nsig, 1);
            S = dz' * (obj.Wc' .* dz) + obj.R;
            dx = sig - repmat(obj.x', nsig, 1);
            Pxz = dx' * (obj.Wc' .* dz);             % n x m
            K = (S \ Pxz')';                         % K = Pxz / S
            obj.x = obj.x + K * (z - zm');
            obj.P = obj.P - K * Pxz';
            obj.P = (obj.P + obj.P') / 2;
        end
    end
end

File: python/kalman/ukf.py
"""Unscented Kalman filter (numpy port of the C ``kf_ukf_*`` API).

Nonlinear models without Jacobians: the distribution is sampled at 2n+1 sigma
points (from the Cholesky factor of P), propagated through f/h, and the mean
and covariance reconstructed from the results.

Sigma-point parameters: alpha (spread), beta (distribution prior), kappa
(secondary scaling). Defaults alpha=1, beta=2, kappa=0 are robust
(all-positive weights).
"""

from __future__ import annotations

import numpy as np

from .matrix import symmetrize


class UnscentedKalmanFilter:
    """Unscented Kalman filter.

    Parameters
    ----------
    n, m : int
        State / measurement dimension.
    f : callable(x, u, dt) -> x_out
        State-transition function.
    h : callable(x) -> z_out
        Measurement function.
    """

    def __init__(self, n, m, f=None, h=None):
        self.n = int(n)
        self.m = int(m)
        self.x = np.zeros(self.n)
        self.P = np.eye(self.n)
        self.Q = np.zeros((self.n, self.n))
        self.R = np.zeros((self.m, self.m))

        self.f = f
        self.h = h

        self.alpha = 1.0
        self.beta = 2.0
        self.kappa = 0.0
        # lambda = alpha^2 (n + kappa) - n  = 0 for the defaults (matches C)
        self.lambda_ = self.alpha * self.alpha * (self.n + self.kappa) - self.n
        self._recompute_weights()

    def set_models(self, f, h) -> None:
        self.f, self.h = f, h

    def set_parameters(self, alpha, beta, kappa) -> None:
        if alpha <= 0 or beta < 0:
            raise ValueError("alpha must be > 0 and beta >= 0")
        self.alpha = float(alpha)
        self.beta = float(beta)
        self.kappa = float(kappa)
        self.lambda_ = alpha**2 * (self.n + kappa) - self.n
        self._recompute_weights()

    def _recompute_weights(self) -> None:
        n, lam = self.n, self.lambda_
        denom = n + lam
        nsig = 2 * n + 1
        self.Wm = np.full(nsig, 0.5 / denom)
        self.Wc = np.full(nsig, 0.5 / denom)
        self.Wm[0] = lam / denom
        self.Wc[0] = lam / denom + (1.0 - self.alpha**2 + self.beta)

    # --- state / covariance / noise ---
    def set_state(self, x) -> None:
        self.x[:] = np.asarray(x, dtype=float)

    def get_state(self) -> np.ndarray:
        return self.x.copy()

    def set_covariance(self, P) -> None:
        self.P[:] = np.asarray(P, dtype=float)

    def set_covariance_diagonal(self, diag) -> None:
        self.P[:] = np.diag(np.asarray(diag, dtype=float))

    def set_covariance_scalar(self, p) -> None:
        self.P[:] = float(p) * np.eye(self.n)

    def get_covariance(self) -> np.ndarray:
        return self.P.copy()

    def set_process_noise(self, Q) -> None:
        self.Q[:] = np.asarray(Q, dtype=float)

    def set_process_noise_scalar(self, q) -> None:
        self.Q[:] = float(q) * np.eye(self.n)

    def set_measurement_noise(self, R) -> None:
        self.R[:] = np.asarray(R, dtype=float)

    def set_measurement_noise_scalar(self, r) -> None:
        self.R[:] = float(r) * np.eye(self.m)

    # --- sigma points ---
    def _sigma_points(self):
        n = self.n
        L = np.linalg.cholesky(self.P)      # lower; P = L L^T
        c = np.sqrt(n + self.lambda_)
        sig = np.empty((2 * n + 1, n))
        sig[0] = self.x
        for i in range(n):
            sig[i + 1] = self.x + c * L[:, i]
            sig[i + 1 + n] = self.x - c * L[:, i]
        return sig

    # --- filtering ---
    def predict(self, u=None, dt=0.0) -> None:
        if self.f is None:
            raise RuntimeError("model f not set; call set_models() first")
        sig = self._sigma_points()
        X = np.array([self.f(si, u, dt) for si in sig])
        xm = self.Wm @ X
        d = X - xm
        self.P[:] = symmetrize(d.T @ (self.Wc[:, None] * d) + self.Q)
        self.x[:] = xm

    def update(self, z) -> None:
        if self.h is None:
            raise RuntimeError("model h not set; call set_models() first")
        z = np.asarray(z, dtype=float)
        sig = self._sigma_points()
        Z = np.array([self.h(si) for si in sig])
        zm = self.Wm @ Z

        dz = Z - zm
        S = dz.T @ (self.Wc[:, None] * dz) + self.R

        dx = sig - self.x
        Pxz = dx.T @ (self.Wc[:, None] * dz)     # n x m

        K = Pxz @ np.linalg.inv(S)               # n x m
        self.x = self.x + K @ (z - zm)
        self.P[:] = symmetrize(self.P - K @ Pxz.T)
